### PR TITLE
Small changes

### PR DESCRIPTION
**Pull request by ostcar**
This are some changes, that (I think) make the code more idiomatic.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -16,6 +17,7 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
+// Route is ...
 type Route struct {
```
Comment (ostcar) [inline server.go:21]: An uppercase type is exported. go-lint demands, that it has a docstring. I was not able to find a good one :)

File: server.go [hunk 2]
```diff
@@ -24,113 +26,140 @@ type Route struct {
...
 
-func (b ByPathLength) Len() int           { return len(b) }
-func (b ByPathLength) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
-func (b ByPathLength) Less(i, j int) bool { return len(b[i].Path) > len(b[j].Path) }
+func (b byPathLength) Len() int           { return len(b) }
+func (b byPathLength) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
+func (b byPathLength) Less(i, j int) bool { return len(b[i].Path) > len(b[j].Path) }
 
-const TIMEOUT_MILLISECONDS = 500
+const timeout = 500 * time.Millisecond
```
Comment (ostcar) [inline server.go:40]: In go, a duration value can be constant. If you do not put the type in the definition, it gets more flexible. Now, it is easy to set another unit like 

```go
const timeout = 3 * time.Second
```

File: server.go [hunk 3]
```diff
@@ -24,113 +26,140 @@ type Route struct {
...
 func main() {
+	if err := run(); err != nil {
+		log.Fatalf("Error: %v\n", err)
+	}
+}
 
-	cgi_scripts_dir := os.Getenv("CGI_DIR")
-	if cgi_scripts_dir == "" {
-		log.Fatalf("ERROR: CGI_DIR environment variable not set\n")
+func run() error {
```
Comment (ostcar) [inline server.go:50]: The go error handling is a bit strange. One idea is, to return the error as far as possible and only log it at one place. A `run() error` function helps with that.

File: server.go [hunk 4]
```diff
@@ -24,113 +26,140 @@ type Route struct {
...
 		}
-
 	}
 
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-
 		start := time.Now()
 
-		handleHTTPRequest(w, r, routes, cgi_scripts_dir)
+		if err := handleHTTPRequest(w, r, routes, cgiScriptsDir); err != nil {
```
Comment (ostcar) [inline server.go:82]: Here is the same. When the handle function returns an error, than it can be handled at one place.

File: server.go [hunk 5]
```diff
@@ -24,113 +26,140 @@ type Route struct {
...
 
-	sort.Sort(ByPathLength(routes.Routes))
+	sort.Sort(byPathLength(routes.Routes))
 
-	return routes
+	return routes, nil
 }
 
-func findRoute(r *http.Request, routes Routes) *Route {
+func findRoute(r *http.Request, routes Routes) (*Route, bool) {
```
Comment (ostcar) [inline server.go:124]: In a lot of places, go uses a second return argument to return "success". To return `nil` does also work, but to add a second return value gives you more flexibility, for example return `(nil, true)`

File: server.go [hunk 6]
```diff
@@ -24,113 +26,140 @@ type Route struct {
...
 
-func handleHTTPRequest(w http.ResponseWriter, r *http.Request, routes Routes, cgi_scripts_dir string) {
-	route := findRoute(r, routes)
-	if route == nil {
-		http.Error(w, "no script found for route", http.StatusBadRequest)
-		return
+func handleHTTPRequest(w http.ResponseWriter, r *http.Request, routes Routes, cgiScriptsDir string) error {
+	route, ok := findRoute(r, routes)
+	if !ok {
+		return errors.Join(
```
Comment (ostcar) [inline server.go:139]: errors.Join was added in go 1.20 (current release) If you want to support older versions of go, another solution has to be used.

File: server.go [hunk 7]
```diff
@@ -24,113 +26,140 @@ type Route struct {
...
+	return fmt.Sprintf("status code: %d", err)
+}
 
-	ctx, cancel := context.WithTimeout(context.Background(), TIMEOUT_MILLISECONDS*time.Millisecond)
+func executeScript(w http.ResponseWriter, r *http.Request, route Route, binaryPath string) error {
+	ctx, cancel := context.WithTimeout(r.Context(), timeout)
 	defer cancel()
 
-	cmd := exec.CommandContext(ctx, binary_path)
+	cmd := exec.CommandContext(ctx, binaryPath)
```
Comment (ostcar) [inline server.go:162]: I have no experience with the exec package. But the documentation says, that `exec.CommandContext` will stop the process, if the context get done. So there is no need to manually handle that.

There was a timing bug in your code. When there were many parallel requests, the context could timeout before the process had even started. In this case `cmd.Process` in your line 166 was nil and calling `cmd.Process.Kill()` resulted in a panic.

File: server.go [hunk 8]
```diff
@@ -183,13 +204,16 @@ func parseURLParameters(url, pattern string) map[string]string {
 		return nil
 	}
 
-	parameters := make(map[string]string)
+	parameters := make(map[string]string, len(patternSegments))
```
Comment (ostcar) [inline server.go:207]: The len argument is optional. It helps the runtime to initialize the map with a good size.

File: server.go [hunk 9]
```diff
@@ -183,13 +204,16 @@ func parseURLParameters(url, pattern string) map[string]string {
...
 
 	for i, segment := range patternSegments {
 		if len(segment) > 2 && segment[0] == '{' && segment[len(segment)-1] == '}' {
 			paramName := segment[1 : len(segment)-1]
 			parameters[paramName] = urlSegments[i]
-		} else if segment != urlSegments[i] {
+			continue
+		}
+
+		if segment != urlSegments[i] {
```
Comment (ostcar) [inline server.go:216]: Some go developer say, that you should not use `else` in go. They try to return early.

**Review by lukewilliamboswell — APPROVED** [review]
Thank you @ostcar, this has taken me a while to get to. Very much appreciated.

**Pull request merged**